mobile_verifier/src/promotion_reward.rs

```diff
@@ -0,0 +1,644 @@
+            "#,
+        )
+        .bind(epoch.start)
+        .bind(epoch.end)
+        .fetch(pool)
+        .map_err(anyhow::Error::from)
+        .and_then(|x: PromotionRewardShares| async move {
+            let service_provider_id = carrier
+                .payer_key_to_service_provider(&x.carrier_key.to_string())
+                .await? as ServiceProviderId;
+            Ok(ServiceProviderPromotionRewardShares {
+                service_provider_id,
+                rewardable_entity: x.rewardable_entity,
+                shares: x.shares,
+            })
+        })
+        .try_collect()
+        .await?;
+        Ok(Self { rewards })
+    }
+
+    pub fn into_rewards<'a>(
+        self,
+        sp_rewards: &mut ServiceProviderRewards,
+        unallocated_sp_rewards: Decimal,
+        reward_period: &'a Range<DateTime<Utc>>,
+    ) -> impl Iterator<Item = (u64, proto::MobileRewardShare)> + 'a {
+        let total_promotion_rewards_allocated =
+            sp_rewards.get_total_rewards_allocated_for_promotion();
+        let total_shares_per_service_provider = self.rewards.iter().fold(
+            HashMap::<i32, Decimal>::default(),
+            |mut shares, promotion_reward_share| {
+                *shares
+                    .entry(promotion_reward_share.service_provider_id)
+                    .or_default() += Decimal::from(promotion_reward_share.shares);
+                shares
+            },
+        );
+        let sp_promotion_reward_shares: HashMap<_, _> = total_shares_per_service_provider
+            .iter()
+            .map(|(sp, total_shares)| {
+                if total_promotion_rewards_allocated.is_zero() || total_shares.is_zero() {
+                    (*sp, SpPromotionRewardShares::default())
+                } else {
+                    let rewards_allocated_for_promotion =
+                        sp_rewards.take_rewards_allocated_for_promotion(sp);
+                    let share_of_unallocated_pool =
+                        rewards_allocated_for_promotion / total_promotion_rewards_allocated;
+                    let sp_share = SpPromotionRewardShares {
+                        shares_per_reward: rewards_allocated_for_promotion / total_shares,
+                        shares_per_matched_reward: unallocated_sp_rewards
+                            * share_of_unallocated_pool
+                            / total_shares,
+                    };
+                    (*sp, sp_share)
+                }
+            })
+            .collect();
+
+        self.rewards
+            .into_iter()
+            .map(move |reward| {
+                let shares = Decimal::from(reward.shares);
+                let sp_promotion_reward_share = sp_promotion_reward_shares
+                    .get(&reward.service_provider_id)
+                    .copied()
+                    .unwrap_or_default();
+                let service_provider_amount = sp_promotion_reward_share.shares_per_reward * shares;
```

Comment:
```suggestion
                let service_provider_token = sp_promotion_reward_share.shares_per_reward * shares;
```
Should this be token seems more clear than just `amount`?

Reply:
This looks like to be Mobile Bones.